dlib/function.h

```diff
@@ -0,0 +1,91 @@
+// Copyright (C) 2022  Davis E. King (davis@dlib.net)
+// License: Boost Software License   See LICENSE.txt for the full license.
+#ifndef DLIB_FUNCTION_H_
+#define DLIB_FUNCTION_H_
+
+#include <cstddef>
+#include "te.h"
+#include "invoke.h"
+
+namespace dlib
+{
+    template <
+        class Storage, 
+        class F
+    > 
+    class function_basic;
+
+    template <
+        class Storage,
+        class R, 
+        class... Args
+    > 
+    class function_basic<Storage, R(Args...)> 
+    {
+    public:
+        constexpr function_basic(std::nullptr_t) {}
+        constexpr function_basic()                                          = default;
+        constexpr function_basic(const function_basic& other)               = default;
+        constexpr function_basic& operator=(const function_basic& other)    = default;
+        constexpr function_basic(function_basic&& other)                    = default;        
+        constexpr function_basic& operator=(function_basic&& other)         = default;
+
+        template <
+            typename F,
+            std::enable_if_t<!std::is_same<std::decay_t<F>, function_basic>::value &&
+                             dlib::is_invocable_r<R, F&&, Args...>::value,
+                             bool> = true
+        >
+        function_basic(
+            F&& f
+        ) : str{std::forward<F>(f)},
+            func{[](void* self, Args... args) -> R {
+                return dlib::invoke(*reinterpret_cast<std::add_pointer_t<F>>(self),
+                                    std::forward<Args>(args)...);
+            }}
+        {
+        }
+
+        template <
+            typename F,
+            std::enable_if_t<!std::is_same<std::decay_t<F>, function_basic>::value &&
+                             dlib::is_invocable_r<R, F&&, Args...>::value,
+                             bool> = true
+        >
+        function_basic& operator=(F&& f)
+        {
+            *this = std::move(function_basic{std::forward<F>(f)});
+            return *this;
+        }
+
+        explicit operator bool() const noexcept
+        {
+            return !str.is_empty() && func;
+        }
+
+        R operator()(Args... args) const {
+            return func.ptr(const_cast<void*>(str.get_ptr()), std::forward<Args>(args)...);
+        }
+
+    private:
+        Storage str;
+        te::managed_move_pointer<R(void*, Args...), true> func;
+    };
+
+    template <class F> 
+    using function_heap  = function_basic<te::storage_heap, F>;
```

Comment:
These are all nice.  We should replace the `dlib::any_function` code though.  That thing is so aggressively pre-c++11.  Like let's just put this stuff in any_function.h and name the `function_heap` one `any_function`.  That would be super great :D

Reply:
Yeah cool! I gave up reading `dlib::any_function` and didn't attempt to change it. But it's doing the same thing as `std::function` or `function_heap` then great let's just refactor to that.

Reply:
Yeah just make it have the same member functions so it's backwards compatible.  Then all that preprocessor magic used to get a pre-c++-11 kinda variadic template goes away.